Scripts/Python/xRegisterAge.py

```diff
@@ -0,0 +1,120 @@
+from PlasmaTypes import *
+from PlasmaKITypes import *
+
+act = ptAttribActivator(1, "Act: Buttons")
+resp = ptAttribResponder(2, "Resp: Button Push")
+ageFileName = ptAttribString(3, "age filename")
+ageInstanceName = ptAttribString(4, "age instance name")
+
+class xRegisterAge(ptResponder):
+
+    def __init__(self):
+        # run parent class init
+        ptResponder.__init__(self)
+        self.id = 5008
+        version = 1
+        self.version = version
+        PtDebugPrint("__init__xRegisterAge v.", version)
+        
+    def OnNotify(self,state,id,events):
+        if state and PtFindAvatar(events) == PtGetLocalAvatar():
+            if id==act.id:
+                if not self.IFindAgeLinkInFolder(ageFileName.value):
+                    if resp.value:
+                        resp.run(self.key,events=events)
+                    PtSendKIMessage(kKILocalChatStatusMsg, PtGetLocalizedString("Global.Messages.AgeBookGet", [ageInstanceName.value]))
+                    self.CheckAndRegisterAge(ageFileName.value, ageInstanceName.value)
+
+    def IFindAgeLinkInFolder(self, ageName):
+        vault = ptVault()
+        folder = vault.getAgesIOwnFolder()
+        ageName = ageName.lower()
+        contents = folder.getChildNodeRefList()
+        for content in contents:
+            link = content.getChild().upcastToAgeLinkNode()
+            if link is not None:
+                info = link.getAgeInfo()
+            else:
+                link = content.getChild()
+                info = link.upcastToAgeInfoNode()
+
+            if info.getAgeFilename().lower() == ageName:
+                return link
+        return None
+
+    def CheckAndRegisterAge(self, ageFileName, ageInstanceName):
+        PtDebugPrint("xRegisterAge.CheckAndRegisterAge(): ",ageFileName,"; ",ageInstanceName)
+        vault = ptVault()
+        ageStruct = ptAgeInfoStruct()
+        ageStruct.setAgeFilename(ageFileName)
+        ageLinkNode = vault.getOwnedAgeLink(ageStruct)
+        if not ageLinkNode:
+            info = ptAgeInfoStruct()
+            info.setAgeFilename(ageFileName)
+            info.setAgeInstanceName(ageInstanceName)
+
+            playerName = PtGetClientName()
+            ageGuid = PtGuidGenerate()
+            userDefName = ""
+            desc = ""
+
+            if playerName[-1] == "s" or playerName[-1] == "S":
+                userDefName = "%s'" % playerName
+                desc = "%s' %s" % (playerName, info.getAgeInstanceName())
+            else:
+                userDefName = "%s's" % playerName
+                desc = "%s's %s" % (playerName, info.getAgeInstanceName())
+
+            info.setAgeInstanceGuid(ageGuid)
+            info.setAgeUserDefinedName(userDefName)
+            info.setAgeDescription(desc)
+
+            link = ptAgeLinkStruct()
+            link.setAgeInfo(info)
+
+            ptVault().registerOwnedAge(link)
+            PtDebugPrint("xRegisterAge.CheckAndRegisterAge(): Registered age - ", ageFileName)
```

Comment:
- The owned Age check is run three times in the OP. Once brute forced to figure out if the message should be shown (and proceeding), once before registering the Age in the Python, and again before registering the Age in the C++. Reduce to once for showing the message (and proceeding) and the C++ side check.
- No need to spec a guid. The server will generate one for us.
- Use f-strings.
- No need to use `xLocTools`. It looks like the assumption is that English gets stored in the vault and `xLocTools` massages that.

```suggestion
    def OnNotify(self, state, id, events):
        if PtWasLocallyNotified(self.ley) and state and id == act.id:
            if not self.IHaveAgeLink(ageFileName.value):
                if resp.value:
                    resp.run(self.key, events=events)
                self.IRegisterAge(ageFileName.value, ageInstanceName.value)
                PtSendKIMessage(kKILocalChatStatusMsg, PtGetLocalizedString("Global.Messages.AgeBookGet", [ageInstanceName.value]))

    def IHaveAgeLink(self, ageName: str) -> bool:
        vault = ptVault()
        info = ptAgeInfoStruct()
        ais.setAgeFilename(ageName)
        return vault.getOwnedAgeLink(ais) is not None

    def IRegisterAge(self, ageFileName: str, ageInstanceName: str):
        PtDebugPrint(f"xRegisterAge.CheckAndRegisterAge(): {ageFileName=} {ageInstanceName=}", level=kWarningLevel)
        if not ageLinkNode:
            playerName = PtGetClientName()
            userDefName = f"{playerName}'" if playerName[-1] in {"s", "S"} else f"{playerName}'s" 
            desc = f"{userDefName} {ageInstanceName}"

            info = ptAgeInfoStruct()
            info.setAgeFilename(ageFileName)
            info.setAgeInstanceName(ageInstanceName)
            info.setAgeUserDefinedName(userDefName)
            info.setAgeDescription(desc)

            link = ptAgeLinkStruct()
            link.setAgeInfo(info)

            ptVault().registerOwnedAge(link)
            PtDebugPrint(f"xRegisterAge.CheckAndRegisterAge(): Registered age {ageFileName=}", level=kWarningLevel)
 ```